state/protocol/badger/state.go

```diff
@@ -14,6 +14,29 @@ import (
 	"github.com/onflow/flow-go/storage/badger/operation"
 )
 
+type MutatorFactory interface {
```

Comment:
I'm not sure it's necessary to expose this outside of the `protocol` package. Creating a `Mutator` is an internal detail of the `protocol state` component, I don't think we would want or need to inject different implementations. IMO it would be simpler to add `results storage.ExecutionResults` to the `NewState` constructor instead and handle this factory logic internally in `Mutate()` or a helper function.

Reply:
We discussed this factory on the meeting, basically it's main purpose is to introduce extra data to the `Mutator` without adding extra fields to `State`. So state stays free of extra fields and new instances of `Mutator` can be injected using `MutatorFactory`

Reply:
@jordanschalm and I discussed this earlier today.  

This implementation here could be much simpler, if we addressed a more conceptual separation-of-concern  problem:
* When mutating a state, we need to verify the validity of the entries in the payload; currently, it is only Receipts, but there will be several more in the future. 
* For verifying an Entity, we need access to the `protocol.ReadOnlyState` 

#### Suggestion 

For Yurii's PR into `master`, I think we should **stick to the Factory pattern**, the way Yurii has currently implemented it. 

I'll attempt a refactoring of the Protocol State APIs beginning of next week (-> created [issue #5103](https://github.com/dapperlabs/flow-go/issues/5103)) and put up a draft PR.